build-logic/cleanup/src/main/java/gradlebuild/cleanup/KillLeakingJavaProcesses.java

```diff
@@ -0,0 +1,248 @@
+ *
+ * Usage: java build-logic/cleanup/src/main/java/gradlebuild/cleanup/KillLeakingJavaProcesses.java
+ *
+ * Kill all Java processes matching:
+ * 1. Main class is `GradleWorkerMain`/`KotlinCompileDaemon`.
+ * 2. Main class is `GradleDaemon` and classpath is inside the current working directory.
+ */
+public class KillLeakingJavaProcesses {
+    private static final Pattern UNIX_JAVA_PROCESS_PATTERN = Pattern.compile("^(\\d+)\\s+(.*/bin/java)\\s+(.*)$");
+    private static final Pattern WINDOWS_JAVA_PROCESS_PATTERN = Pattern.compile("^\"?(.*[/\\\\]bin[/\\\\]java(\\.exe)?)\"\\s+(.*)\\s+(\\d+)$");
+    private static final Pattern MAIN_CLASS_PATTERN = Pattern.compile("([a-z]+\\.)+[A-Z]\\w+");
+
+    private static final String WORKING_DIRECTORY_PATH = new File("").getAbsolutePath();
+
+    public static void main(String[] args) {
+        System.out.println("Working dir is: " + WORKING_DIRECTORY_PATH);
+        ps().stream()
+            .peek(javaProcessInfo -> System.out.println("Found java process " + javaProcessInfo))
+            .filter(JavaProcessInfo::isLeakingJavaProcess)
+            .forEach(JavaProcessInfo::kill);
+
+        System.out.println("After cleanup:");
+        System.out.println(run(determinePsCommand()));
+    }
+
+    private static List<JavaProcessInfo> ps() {
+        ExecResult result = run(determinePsCommand()).assertZeroExit();
+        return JavaProcessInfo.parsePsOutput(result.stdout, isWindows());
+    }
+
+    private static String[] determinePsCommand() {
+        if (isWindows()) {
+            return new String[]{"wmic", "process", "get", "processid,commandline"};
+        } else if (isMacOS()) {
+            return new String[]{"ps", "x", "-o", "pid,command"};
+        } else {
+            return new String[]{"ps", "x", "-o", "pid,cmd"};
+        }
+    }
+
+    static class JavaProcessInfo {
+        /**
+         * The pid of the process.
+         */
+        final String pid;
+        /**
+         * The command line arguments of the process. The first argument is java executable.
+         */
+        final List<String> commands = new ArrayList<>();
+        /**
+         * The main class of the process. May be null.
+         */
+        String mainClass;
+        /**
+         * The classpath passed to the process via `-cp` or `-classpath`. May be null.
+         */
+        String classpath;
+
+        private JavaProcessInfo(String pid, String javaCommand, String otherCommands) {
+            this.pid = pid;
+            commands.add(javaCommand);
+            Stream.of(otherCommands.split("\\s"))
+                .map(String::trim)
+                .filter(s -> !s.isEmpty())
+                .forEach(commands::add);
+            for (int i = 0; i < commands.size(); ++i) {
+                String command = commands.get(i);
+                if (MAIN_CLASS_PATTERN.matcher(command).matches()) {
```

Comment:
I suggest we only do this if we didn't find a `mainClass`, yet.

Reply:
Not an issue any more.